File: wallet/ssi/src/main/kotlin/com/tcc/wallet/ssi/IssuerTrust.kt
package com.tcc.wallet.ssi

import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.crypto.ECDSAVerifier
import com.nimbusds.jose.jwk.Curve
import com.nimbusds.jose.jwk.ECKey
import com.nimbusds.jwt.SignedJWT
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.net.URI
import java.security.cert.CertPathValidator
import java.security.cert.CertificateFactory
import java.security.cert.PKIXParameters
import java.security.cert.TrustAnchor
import java.security.cert.X509Certificate
import java.security.interfaces.ECPublicKey
import java.time.Instant
import java.util.Base64
import java.util.Date
import javax.naming.ldap.LdapName

/**
 * **Issuer-credential validation at OID4VCI receipt time** — the wallet's side of the
 * HAIP §6.1.1 `x5c` issuer-trust model. Mirrors the verifier's reference implementation
 * `ssi_core::x509` (chain validation, leaf-not-CA, `iss`↔SAN binding) so the wallet and
 * verifier agree on what a trustworthy issuer credential is.
 *
 * Before a freshly-received credential is stored, [verifyCredential]:
 *  1. parses the `x5c` JOSE header of the issuer-signed SD-JWT (the part before the first `~`);
 *  2. resolves the issuer signing key from the **leaf** certificate (`x5c[0]`) and verifies the
 *     issuer JWS **ES256** signature over the credential — rejecting a forged/tampered credential;
 *  3. validates the certificate **chain** (leaf → intermediate …) up to a trusted **root anchor**
 *     held locally (HAIP: the leaf MUST NOT be self-signed/CA, and the root MUST NOT be in `x5c`);
 *  4. binds identity: the SD-JWT `iss` claim MUST match the leaf certificate (SAN / CN), exactly
 *     as the verifier does.
 *
 * Any failure throws, so an invalid credential is never returned to be stored. The wallet still
 * stores and later forwards the issuer JWT **unchanged** — this only gates *acceptance*.
 *
 * Trust anchor: the bundled mock **ICP-Brasil** root (`/trust/icp_brasil_root.pem`, a copy of the
 * verifier's default anchor `ssi_core::trust::ICP_BRASIL_MOCK_ROOT_PEM`).
 *
 * Chain + certificate handling uses the JVM standard library
 * (`CertificateFactory`/`CertPathValidator`/`PKIXParameters`); ES256 JWS verification reuses the
 * Nimbus pattern already used for JAR requests ([Jar]).
 */
class IssuerTrust(
    private val anchors: List<X509Certificate>,
    private val enabled: Boolean = true,
    private val clock: () -> Instant = Instant::now,
) {

    /** Validate a received compact SD-JWT VC; throws if the issuer credential is not trustworthy. */
    fun verifyCredential(sdJwt: String) {
        if (!enabled) return

        // The issuer-signed JWT is everything before the first '~' (the SD-JWT VC issuer part).
        val issuerJwt = sdJwt.substringBefore('~')
        verifyX5cJwt(issuerJwt)
    }

    /**
     * Verify **signed Credential Issuer Metadata** (OID4VCI §11.2.3, HAIP §4.1 / VCI-8): the
     * `signed_metadata` JWT served in `/.well-known/openid-credential-issuer`. Validates the JWS
     * with exactly the same HAIP §6.1.1 machinery as a credential — `x5c` chain → trusted anchor,
     * ES256 signature under the leaf, leaf-not-CA / no self-signed cert in `x5c`, and `iss`↔leaf
     * binding — and additionally binds the document to [expectedIssuer], the Credential Issuer
     * Identifier the wallet is talking to (the metadata's `sub` / `credential_issuer` claim per
     * §11.2.3). Returns the verified metadata claims; throws if anything fails.
     */
    fun verifySignedMetadata(signedMetadata: String, expectedIssuer: String): JSONObject {
        if (!enabled) return Jws.decodeUnverified(signedMetadata).second

        val payload = verifyX5cJwt(signedMetadata)

        // Bind the signed metadata to the issuer being talked to. §11.2.3 sets `sub` to the
        // Credential Issuer Identifier; we also accept the mirrored `credential_issuer` claim.
        val sub = payload.optString("sub")
        val credentialIssuer = payload.optString("credential_issuer")
        require(sub == expectedIssuer || credentialIssuer == expectedIssuer) {
            "signed metadata does not authenticate the expected issuer '$expectedIssuer' " +
                "(sub='$sub', credential_issuer='$credentialIssuer')"
        }
        return payload
    }

    /**
     * Verify a compact `x5c`-signed JWS under the HAIP §6.1.1 issuer-trust model and return its
     * payload. Shared by [verifyCredential] (the issuer-signed SD-JWT) and [verifySignedMetadata]
     * (the OID4VCI §11.2.3 `signed_metadata` JWT): both resolve the signing key from the leaf of
     * the `x5c` chain, enforce the HAIP structural constraints, validate the chain to a trusted
     * anchor, check the ES256 signature, and bind the JWT's `iss` to the leaf certificate.
     */
    private fun verifyX5cJwt(jwt: String): JSONObject {
        require(jwt.count { it == '.' } == 2) { "value is not a compact JWS" }
        val (header, payload) = Jws.decodeUnverified(jwt)

        // 1. x5c header → certificate chain (leaf first).
        val x5c = header.optJSONArray("x5c")
            ?: throw IllegalStateException("JWS header carries no x5c (HAIP §6.1.1)")
        val chain = parseX5c(x5c)
        val leaf = chain.first()

        // 2. HAIP structural constraints on the presented chain.
        //    The signing certificate is an end-entity cert: it MUST NOT assert cA.
        require(leaf.basicConstraints < 0) {
            "leaf certificate is a CA, not an end-entity signer (HAIP §6.1.1)"
        }
        //    The trust-anchor root is held locally and MUST NOT be present in x5c, so the chain
        //    MUST NOT contain a self-signed certificate (a root is self-signed).
        require(chain.none { it.subjectX500Principal == it.issuerX500Principal }) {
            "x5c must not contain a self-signed (root) certificate; the anchor is held locally (HAIP §6.1.1)"
        }

        // 3. Chain validation up to a trusted anchor (validity, cA on issuers, links, anchor).
        validateChain(chain)

        // 4. JWS signature under the leaf certificate's key (ES256).
        verifyIssuerSignature(jwt, leaf)

        // 5. Bind the JWT's iss to the leaf certificate (string compare; never dereferenced).
        val iss = payload.optString("iss")
        require(iss.isNotBlank()) { "JWS has no iss claim to bind to the leaf certificate" }
        requireIssMatchesLeaf(iss, leaf)
        return payload
    }

    /** Parse `x5c` (a JSON array of **standard** base64 DER strings, leaf first — RFC 7515 §4.1.6). */
    private fun parseX5c(x5c: JSONArray): List<X509Certificate> {
        require(x5c.length() > 0) { "x5c header is empty" }
        val cf = CertificateFactory.getInstance("X.509")
        return (0 until x5c.length()).map { i ->
            val der = Base64.getMimeDecoder().decode(x5c.getString(i))
            cf.generateCertificate(ByteArrayInputStream(der)) as X509Certificate
        }
    }

    /** PKIX validation of `[leaf, intermediate, …]` (root excluded) up to a trusted anchor. */
    private fun validateChain(chain: List<X509Certificate>) {
        require(anchors.isNotEmpty()) { "no trusted anchors configured" }
        val certPath = CertificateFactory.getInstance("X.509").generateCertPath(chain)
        val anchorSet = anchors.map { TrustAnchor(it, null) }.toSet()
        val params = PKIXParameters(anchorSet).apply {
            isRevocationEnabled = false // no CRL/OCSP for the mock prototype PKI
            date = Date.from(clock())
        }
        try {
            CertPathValidator.getInstance("PKIX").validate(certPath, params)
        } catch (e: Exception) {
            throw IllegalStateException(
                "issuer certificate chain does not validate to a trusted anchor: ${e.message}",
                e,
            )
        }
    }

    /** Verify the issuer JWS over [issuerJwt] under the [leaf]'s public key (must be ES256). */
    private fun verifyIssuerSignature(issuerJwt: String, leaf: X509Certificate) {
        // Build the ECKey from the leaf's public key directly (avoids ECKey.parse(cert), which
        // pulls in bcpkix; only bcprov is on the classpath). The mock PKI is P-256 throughout.
        val ecPub = leaf.publicKey as? ECPublicKey
            ?: throw IllegalStateException("leaf certificate key is not EC (P-256/ES256 expected)")
        val verifier = ECDSAVerifier(ECKey.Builder(Curve.P_256, ecPub).build())
        val jwt = SignedJWT.parse(issuerJwt)
        require(jwt.header.algorithm == JWSAlgorithm.ES256) { "issuer JWS must be signed with ES256" }
        require(jwt.verify(verifier)) {
            "issuer JWS signature does not verify under the leaf certificate key"
        }
    }

    /** Bind `iss` to the leaf certificate via SAN (dNSName/URI) or CN — never dereferenced. */
    private fun requireIssMatchesLeaf(iss: String, leaf: X509Certificate) {
        val sans = leafSans(leaf)
        if (iss in sans) return
        val cn = subjectCn(leaf)
        val host = runCatching { URI(iss).host }.getOrNull()
        if (host != null && (host in sans || host == cn)) return
        if (cn == iss) return
        throw IllegalStateException(
            "credential iss '$iss' is not bound to the leaf certificate (SAN=$sans, CN=$cn)",
        )
    }

    /** dNSName (type 2) + uniformResourceIdentifier (type 6) SAN entries. */
    private fun leafSans(leaf: X509Certificate): List<String> =
        leaf.subjectAlternativeNames?.mapNotNull { entry ->
            when (entry[0] as Int) {
                2, 6 -> entry[1] as? String
                else -> null
            }
        } ?: emptyList()

    private fun subjectCn(cert: X509Certificate): String? = runCatching {
        LdapName(cert.subjectX500Principal.name).rdns
            .firstOrNull { it.type.equals("CN", ignoreCase = true) }
            ?.value?.toString()
    }.getOrNull()

    companion object {
        private const val ANCHOR_RESOURCE = "/trust/icp_brasil_root.pem"

        /** The default validator, anchored at the bundled mock ICP-Brasil root. */
        fun default(clock: () -> Instant = Instant::now): IssuerTrust =
            IssuerTrust(bundledAnchors(), clock = clock)

        /** A validator anchored at a caller-supplied set of PEM roots (tests / future config). */
        fun fromPems(pems: List<String>, clock: () -> Instant = Instant::now): IssuerTrust =
            IssuerTrust(pems.map { parsePem(it) }, clock = clock)

        /** A no-op validator — for tests/flows that use stub (unsigned) credentials. */
        fun disabled(): IssuerTrust = IssuerTrust(emptyList(), enabled = false)

        /** The bundled trust anchor(s) from the classpath resource. */
        fun bundledAnchors(): List<X509Certificate> {
            val stream = IssuerTrust::class.java.getResourceAsStream(ANCHOR_RESOURCE)
                ?: error("bundled trust anchor $ANCHOR_RESOURCE missing from the classpath")
            val cf = CertificateFactory.getInstance("X.509")
            return stream.use { cf.generateCertificates(it).map { c -> c as X509Certificate } }
        }

        private fun parsePem(pem: String): X509Certificate =
            CertificateFactory.getInstance("X.509")
                .generateCertificate(ByteArrayInputStream(pem.toByteArray(Charsets.UTF_8))) as X509Certificate
    }
}

File: crates/ssi-core/src/issuer_trust.rs
//! Holder-side issuer acceptance — the wallet's side of HAIP §6.1.1 `x5c` trust.
//!
//! When a wallet receives a credential (or signed Credential Issuer Metadata) it
//! must decide whether the issuer is trustworthy *before* accepting it. That
//! decision is exactly the verifier's `x5c` issuer-trust check — [`crate::oid4vp`]
//! runs the same one at presentation time — so wallet and verifier agree on what a
//! trustworthy issuer credential is. This module bundles those checks behind
//! holder-facing entry points, reusing the shared [`crate::x509`] primitives, and
//! adds the holder-side revocation check that badges a stored credential.

use anyhow::{Result, anyhow, bail};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::resolve::Fetcher;
use crate::trust::TrustStore;
use crate::{crypto, sd_jwt, status, x509};

/// Verify a received compact SD-JWT VC under the HAIP §6.1.1 issuer-trust model:
/// the issuer-signed JWT's `x5c` chain validates to a trusted anchor, its ES256
/// signature verifies under the leaf certificate, and its `iss` is bound to that
/// leaf. The wallet calls this before accepting a credential; the SD-JWT is still
/// stored and later forwarded unchanged (this only gates *acceptance*).
pub fn verify_issuer_credential(sd_jwt: &str, anchors_pem: &[String], now_unix: i64) -> Result<()> {
    let (issuer_jwt, _disclosures) = sd_jwt::split(sd_jwt);
    verify_x5c_jws(&issuer_jwt, anchors_pem, now_unix)?;
    Ok(())
}

/// Verify signed Credential Issuer Metadata (OID4VCI §11.2.3 / HAIP §4.1): the
/// `signed_metadata` JWT validates under the same `x5c` machinery as a credential
/// and is additionally bound to `expected_issuer` — its `sub` / `credential_issuer`
/// claim, the Credential Issuer Identifier the wallet is talking to. Returns the
/// verified metadata claims.
pub fn verify_signed_metadata(
    jwt: &str,
    expected_issuer: &str,
    anchors_pem: &[String],
    now_unix: i64,
) -> Result<Value> {
    let payload = verify_x5c_jws(jwt, anchors_pem, now_unix)?;
    let sub = payload.get("sub").and_then(Value::as_str);
    let credential_issuer = payload.get("credential_issuer").and_then(Value::as_str);
    if sub != Some(expected_issuer) && credential_issuer != Some(expected_issuer) {
        bail!(
            "signed metadata does not authenticate the expected issuer '{expected_issuer}' \
             (sub={sub:?}, credential_issuer={credential_issuer:?})"
        );
    }
    Ok(payload)
}

/// The holder-side revocation state of a stored credential.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CredentialStatus {
    /// The credential carries no status reference — nothing to check.
    Unknown,
    /// The status-list bit is clear: not revoked.
    Fresh,
    /// The status-list bit is set: revoked.
    Revoked,
}

/// Check a stored credential's revocation state against its Token Status List,
/// fetching the list through `fetcher`. Lets the wallet badge revoked credentials.
/// Trust in the status list is established independently of the credential's
/// signing key (its own `x5c` chain → a trusted anchor + `iss` identity binding),
/// so issuer key rotation doesn't break the check.
pub fn credential_status(
    sd_jwt: &str,
    fetcher: &dyn Fetcher,
    anchors_pem: &[String],
    now_unix: i64,
) -> Result<CredentialStatus> {
    let (issuer_jwt, disclosures) = sd_jwt::split(sd_jwt);
    let (_h, issuer_payload) = crypto::decode_jws_unverified(&issuer_jwt)?;
    let claims =
        sd_jwt::reconstruct_claims(&issuer_payload, &disclosures).unwrap_or(issuer_payload);

    let Some((uri, index)) = status::status_reference(&claims) else {
        return Ok(CredentialStatus::Unknown);
    };
    let issuer = claims.get("iss").and_then(Value::as_str);
    let trust = TrustStore::from_pems(anchors_pem)?;
    let jwt = String::from_utf8(fetcher.get(&uri)?)?;
    Ok(match status::check_status(&jwt, index, &trust, now_unix, issuer)? {
        status::StatusCheck::Valid => CredentialStatus::Fresh,
        status::StatusCheck::Revoked => CredentialStatus::Revoked,
    })
}

/// Verify a compact `x5c`-signed JWS under HAIP §6.1.1 and return its payload:
/// resolve the chain from the `x5c` header, validate it to a trusted anchor (which
/// also enforces leaf-not-CA and CA issuers), verify the ES256 signature under the
/// leaf key, and bind the JWT's `iss` to the leaf certificate. Shared by
/// [`verify_issuer_credential`] and [`verify_signed_metadata`], mirroring the
/// verifier's `oid4vp` issuer-trust path so the two never diverge.
fn verify_x5c_jws(jwt: &str, anchors_pem: &[String], now_unix: i64) -> Result<Value> {
    let (header, payload) = crypto::decode_jws_unverified(jwt)?;
    let chain = x509::parse_x5c(
        header
            .get("x5c")
            .ok_or_else(|| anyhow!("JWS header carries no x5c (HAIP §6.1.1)"))?,
    )?;
    let trust = TrustStore::from_pems(anchors_pem)?;

    // Chain → trusted anchor (validity, leaf-not-CA, CA issuers, links, anchor).
    x509::validate_chain(&chain, trust.anchor_certs(), now_unix)
        .map_err(|e| anyhow!("untrusted issuer certificate chain: {e:#}"))?;

    // ES256 signature under the leaf certificate's own key.
    sd_jwt::verify_issuer_signature(jwt, &chain[0].public_jwk()?)?;

    // Bind iss to the leaf certificate (string compare; never dereferenced).
    let iss = payload
        .get("iss")
        .and_then(Value::as_str)
        .ok_or_else(|| anyhow!("JWS has no iss claim to bind to the leaf certificate"))?;
    x509::iss_matches_leaf(iss, &chain[0])?;

    Ok(payload)
}

#[cfg(all(test, feature = "testkit"))]
mod tests {
    use super::*;
    use crate::trust::ICP_BRASIL_MOCK_ROOT_PEM;

    // A fixed "now" inside the 2020..2099 validity window of the mock PKI.
    const NOW: i64 = 1_700_000_000;

    fn anchors() -> Vec<String> {
        vec![ICP_BRASIL_MOCK_ROOT_PEM.to_string()]
    }

    #[test]
    fn accepts_a_genuine_issuer_credential() {
        let demo = crate::testkit::mint(false);
        verify_issuer_credential(&demo.sd_jwt, &anchors(), NOW)
            .expect("the demo diploma chains to the bundled ICP-Brasil root");
    }

    #[test]
    fn rejects_a_credential_under_an_untrusted_anchor() {
        let demo = crate::testkit::mint(false);
        let rogue = vec![include_str!("../fixtures/pki/rogue_root.pem").to_string()];
        assert!(verify_issuer_credential(&demo.sd_jwt, &rogue, NOW).is_err());
    }

    #[test]
    fn reports_fresh_and_revoked_status() {
        let fresh = crate::testkit::mint(false);
        assert_eq!(
            credential_status(&fresh.sd_jwt, &fresh.fetcher, &anchors(), NOW).unwrap(),
            CredentialStatus::Fresh,
        );
        let revoked = crate::testkit::mint(true);
        assert_eq!(
            credential_status(&revoked.sd_jwt, &revoked.fetcher, &anchors(), NOW).unwrap(),
            CredentialStatus::Revoked,
        );
    }

    #[test]
    fn verifies_signed_issuer_metadata_and_rejects_wrong_issuer_or_tampering() {
        use crate::x509::{Cert, demo};
        use p256::ecdsa::SigningKey;
        use p256::pkcs8::DecodePrivateKey;
        use serde_json::json;

        // Sign issuer metadata with the demo leaf key, carrying the leaf+intermediate x5c.
        let leaf_key = SigningKey::from_pkcs8_pem(demo::UFSC_LEAF_KEY_PKCS8_PEM).unwrap();
        let x5c = json!([
            crypto::b64std(Cert::from_pem(demo::UFSC_LEAF_PEM).unwrap().der()),
            crypto::b64std(Cert::from_pem(demo::MEC_INTERMEDIATE_PEM).unwrap().der()),
        ]);
        let issuer = "https://diploma.ufsc.br"; // a SAN of the demo leaf
        let header = json!({ "alg": "ES256", "typ": "openid-credential-issuer+jwt", "x5c": x5c });
        let jwt = crypto::sign_jws_es256(&header, &json!({ "iss": issuer, "sub": issuer }), &leaf_key);

        // Happy path: trusted chain + iss bound to the leaf + sub == the issuer we expect.
        let claims = verify_signed_metadata(&jwt, issuer, &anchors(), NOW).unwrap();
        assert_eq!(claims["sub"], issuer);

        // A document authenticating a different issuer than the one we're talking to is rejected.
        assert!(verify_signed_metadata(&jwt, "https://evil.example", &anchors(), NOW).is_err());

        // A tampered signature is rejected.
        let tampered = format!("{}x", &jwt[..jwt.len() - 1]);
        assert!(verify_signed_metadata(&tampered, issuer, &anchors(), NOW).is_err());
    }
}
